src/Filter/Filter.php

```diff
@@ -141,4 +153,57 @@ protected function getNewParameterName(BaseProxyQueryInterface $query)
         // by underscores.
         return str_replace('.', '_', $this->getName()).'_'.$query->getUniqueParameterId();
     }
+
+    /**
+     * Returns the reference to the `Orx` expression used in the `where` clause.
+     * If it doesn't exist, a new one is created.
+     * This method uses a marker (":sonata_admin_datagrid_filter_query_marker") in
+     * the resulting DQL in order to identify the corresponding "WHERE (...)" condition
+     * group each time it is required.
+     * It allows to get queries like "WHERE previous_condition = previous_value AND (filter_1 = value OR filter_2 = value OR ...)",
+     * where the logical "OR" operators added by the filters are grouped inside a condition,
+     * instead of having unfolded "WHERE ..." clauses like "WHERE previous_condition = previous_value OR filter_1 = value OR filter_2 = value OR ...",
+     * which will produce undesired results.
+     *
+     * NEXT_MAJOR: Remove this method and let the consumers of this class to set
+     * the `Orx` expression using `setOrExpression()`.
```

Comment:
How do you want to do this ?

Reply:
By allowing cases [like this](https://github.com/sonata-project/SonataAdminBundle/blob/6292c61255e945a0aa1440db260680c40de79a14/src/Search/SearchHandler.php#L93) to define the `Xor` instance that will be used within the filter query, by creating a new one or receiving the instance from the first configured filter and passing it to the next in a chained process.
I guess the second option is more feasible since it doesn't require the consumer (the AdminBundle) to know too much details about the implementation (`Sonata\AdminBundle\Datagrid\ProxyQueryInterface` doesn't provide access to the query builder).
I imagine somethig like this:
```php
$orExpression = null;
foreach ($datagrid->getFilters() as $filter) {
    /** @var FilterInterface $filter */
    $formName = $filter->getFormName();

    if ($filter->getOption('global_search', false)) {
        $filter->setCondition(FilterInterface::CONDITION_OR);
        if (null === $orExpression) {
            // let the first used filter to create the `Xor` expression.
            $orExpression = $filter->getOrExpression();
            // or $orExpression = $filter->getOption('or_expression'); to avoid the API change.
        }
        $filter->setOrExpression($orExpression);
        // or $filter->setOption('or_expression', $orExpression);
        $datagrid->setValue($formName, null, $term);
    } elseif (isset($datagridValues[$formName])) {
        // Remove any previously set filter that is not configured for the global search.
        $datagrid->removeFilter($formName);
    }
}
```

Reply:
It would require to add getOrExpression and setOrExpression to the filter interface.
Isn't this too specific to the persistence bundle ?

Reply:
I think no. Please read the comments in the snippet.

Reply:
In your snippet you're still using a `getOrExpression`.
So the `NEXT_MAJOR: Remove this method` comment seems wrong to me ; and the method need to be public

Reply:
Sure. What should be removed in the next major is the workaround related to the marker.
Even we could remove that if we can restrict the constraint for the AdminBundle to a version including the mentioned behavior.

Reply:
>In your snippet you're still using a getOrExpression.

The commented lines are showing a second option based in options to avoid the API change.
To clarify, I'll represent a clear version of that choice here:
```php
$orExpression = null;
foreach ($datagrid->getFilters() as $filter) {
    /** @var FilterInterface $filter */
    $formName = $filter->getFormName();

    if ($filter->getOption('global_search', false)) {
        $filter->setCondition(FilterInterface::CONDITION_OR);
        if (null === $orExpression) {
            // let the first used filter to create the `Xor` expression.
            $orExpression = $filter->getOption('or_expression');
        } else {
            $filter->setOption('or_expression', $orExpression);
        }
        $datagrid->setValue($formName, null, $term);
    } elseif (isset($datagridValues[$formName])) {
        $datagrid->removeFilter($formName);
    }
}
```
BTW, I'll push a commit with this change soon.

Reply:
@VincentLanglet, I updated this method.
In the last commit I'm using a static array to hold the corresponding `Xor` expression used by each admin. I also left the marker logic as fallback until the constraint for the admin bundle can guarantee that the admin code is provided.